presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java

```diff
@@ -280,24 +283,21 @@ PlanNodeStatsEstimate calculateJoinComplementStats(
                 return PlanNodeStatsEstimate.unknown();
             }
 
-            return leftStats.mapOutputRowCount(rowCount -> 0.0);
+            return normalizer.normalize(leftStats.mapOutputRowCount(rowCount -> 0.0), types);
         }
 
         // TODO: add support for non-equality conditions (e.g: <=, !=, >)
-        int numberOfFilterClauses = filter.map(
-                exression -> ExpressionUtils.extractConjuncts(exression).size())
-                .orElse(0);
+        int numberOfFilterClauses = filter.map(expression -> extractConjuncts(expression).size()).orElse(0);
 
         // Heuristics: select the most selective criteria for join complement clause.
         // Principals behind this heuristics is the same as in computeInnerJoinStats:
         // select "driving join clause" that reduces matched rows the most.
-        return IntStream.range(0, criteria.size())
-                .mapToObj(drivingClauseId -> {
-                    EquiJoinClause drivingClause = criteria.get(drivingClauseId);
-                    return calculateJoinComplementStats(leftStats, rightStats, drivingClause, criteria.size() - 1 + numberOfFilterClauses);
-                })
+        return criteria.stream()
+                .map(drivingClause -> calculateJoinComplementStats(leftStats, rightStats, drivingClause, criteria.size() - 1 + numberOfFilterClauses))
+                .filter(estimate -> !estimate.isOutputRowCountUnknown())
```

Comment:
This seems unrelated to normalizer introduction. Can't normalizer consume unknown stats (I think it should)?

Reply:
Let me extract this into a separate commit. This change favors known stats over unknown stats. Before if any of the stats were unknown, the result was unknown.